### ISSUE TITLE
Feature Proposal: New UI Pagination Control

### ISSUE THREAD

**Issue by SavoySchuler** [Microsoft/microsoft-ui-xaml#60]
# Proposal: New UI Pagination Control

## Summary

This UI pager control provides a user interface for navigating pages of a GridView, ListView, DataGrid and FlipView. 

## Rationale

The absence of a standard UI pager control, last seen as DataPager in Silverlight, has been a pain point in both WPF and UWP that has forced a variety of unfavorable workarounds for developers. Inclusion of this control in UWP would resolve an ecosystem gap and enable developers to efficiently deliver UI paging experiences in their applications. Through XAML Islands, it may also provide an opportunity to benefit WPF developers working with DataGrid or ListView. The scope of this proposal is not provide a data paging solution at this time but to start with UI paging as a foundation on top of which data paging may be integrated.   

## Functional Requirements

| # | Feature | Priority |
|:-:|:--|:-:|
| 1 | Can be used to page GridView, ListView, DataGrid and FlipView.  | Must |
| 2 | Base indexing component display mode can be a button panel or an editable ComboBox.  | Must |
| 3 | Custom text can be added before and after the indexing component. | Must |
| 4 | Able to add text and/or glyph navigation buttons for indexing to the first, last, next, and previous page.  | Must |
| 5 | Able to toggle the state and visibility of navigation buttons when at the end of the indexable range. | Must |
| 7 | Ellipsis can be added to the numerical button panel indexing component to show an ommitted range of indices.  | Must |
| 8 | Designed so that a data paging component may be added. | Must |
| 9 | Numerals in the button panel display mode can be replaced by glyphs. | Should |
| 10 | Can be used to page DataGrid and ListView in WPF through XAML Islands. | Should |

### 1. Using a button panel as the display mode.

![image](https://user-images.githubusercontent.com/16964652/49760313-0fbbac00-fc79-11e8-902a-96a0dd9adab4.png)

```XAML
<UserControl x:Class="DataPagerSample.MainPage">
    <Grid x:Name="LayoutRoot" Width="Auto">
        <StackPanel>
            <GridView x:Name="gridView1" ... />
	        <muxc:UIPager x:Name="uiPager1"
		    Source="{Binding}"
		    NumberOfIndicesShowing="7"
		    EllipsisMaxBefore="5"
		    EllipsisMaxAfter="1" />
        </StackPanel>
    </Grid>
</UserControl>
```
 
### 2. Using an editable ComboBox as the display mode.

![image](https://user-images.githubusercontent.com/16964652/49760326-1813e700-fc79-11e8-974e-8e997796741a.png)

```XAML
<UserControl x:Class="DataPagerSample.MainPage">
    <Grid x:Name="LayoutRoot" Width="Auto">
        <StackPanel>
            <GridView x:Name="gridView1" ... />
	        <muxc:UIPager x:Name="uiPager1"
		    Source="{Binding}"
		    DisplayMode="EditableComboBox"
		    LastButton="None"
		    FirstButton="None"
		    NextButtonText="Prev"
		    PreviousButtonText="Next"
	            PrefixText="Page"  
		    TotalNumberOfPages="10" />
       </StackPanel>
    </Grid>
</UserControl>
```

## Detailed Feature Design

**Visual Components**

 | Component |  Notes |
|:---:|:---|
| DisplayMode | * Used to set either a button panel (default) or an editable ComboBox as the indexing component. <br> * When set to be a button panel, the number of visible indices can be specified. <br><br> ![image](https://user-images.githubusercontent.com/16964652/49760354-23671280-fc79-11e8-804b-71b5b84a80f5.png) &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; ![image](https://user-images.githubusercontent.com/16964652/49760373-2eba3e00-fc79-11e8-8304-12d7b6ee2c89.png) |
| LastButton | * Button displaying text and/or glyph indicating that the user may navigate to the last index. <br> * Automatically disabled when at last index. <br> * Can be set to not be visible when at the last index.  <br><br> ![image](https://user-images.githubusercontent.com/16964652/49760392-3a0d6980-fc79-11e8-8d96-a811cee1c759.png) |
| FirstButton | * Button displaying text and/or glyph indicating that the user may navigate to the first index. <br> * Automatically disabled when at first index. <br> * Can be set to not be visible when at the first index.  <br><br> ![image](https://user-images.githubusercontent.com/16964652/49760409-485b8580-fc79-11e8-9545-a580f6016f6c.png) |
| NextButton | * Button displaying text and/or glyph indicating that the user may navigate to the next index. <br> * Automatically disabled when at last index. <br> * Can be set to not be visible when at the last index. <br><br> ![image](https://user-images.githubusercontent.com/16964652/49760426-54474780-fc79-11e8-9446-ee2886941a24.png) |
| PreviousButton | * Button displaying text and/or glyph indicating that the user may navigate to the previous index. <br> * Automatically disabled when at first index. <br> * Can be set to not be visible when at the first index.  <br><br> ![image](https://user-images.githubusercontent.com/16964652/49760443-62956380-fc79-11e8-8f93-562a92fde666.png) |
| Ellipsis | * Button, often reading "...", used between indexes and before or after the first/last index to indicate an accessible but omitted range of indexes. Clicking this button will index the user to the median of the omitted range.<br> * Clickability can be disabled. <br> * MaxBefore and MaxAfter properties can be used to set  how many indices appear between the current page and the ellipsis before/after after it. <br> * Visibility of the first/last index can be disabled. <br> * Only visbile when using the button panel as the display mode.  <br><br> ![image](https://user-images.githubusercontent.com/16964652/49760513-85277c80-fc79-11e8-926e-3453ea29b0a3.png) |
| PrefixText | * Text displayed before the editable ComboBox indexing component. <br><br> ![image](https://user-images.githubusercontent.com/16964652/49760540-93759880-fc79-11e8-9c4f-3c3222b95a13.png) |
| NumberOfPages | * When a total number of indices (N) is given, this suffix string will appear after the editable ComboBox indexing component and read "of N". Localization will put "N" where it should be in a given language. <br><br> ![image](https://user-images.githubusercontent.com/16964652/49760557-9a9ca680-fc79-11e8-943f-3425d993fdeb.png) |

**Accessibility**

| State | Action | Narrator |
|:---|:---|:---|:---:|
| UI pager is first focused on by tabbing | Focus defaults current page after announcing accessible name of UI pager. | “Page selector. Current page is N." |
| UI pager is tabbed through | Tab Button: <br> Will go through all actionable items in order without regard to groups. <br> <br> Arrow keys: <br> Will be able to explore groups in the specified directions. Pressing the down arrow key at the bottom of the combobox will wrap the user to the top. Pressing the up arrow key at the top of the combobox will wrap the user to the bottom. <br> <br> Escape: <br> Will escape UI pager. <br> <br> Enter and Spacebar: <br> Will select the component focused on. <br> <br> Home: <br> Will move focus to "go back" elements. In the combobox, it will jump the user to the first index. <br> <br> End: <br> Will move focus to "go forward" elements. In the combobox, it will jump the user to the last index. | Narrator will announce an accessible name of the visual component. Ex:<br><br> “first page button” <br><br> “previous page button” <br><br> “1st page” <br><br> “current page” <br><br> "page selection drop down menu: current page is 1” | 

### API Surface Example

```XAML
<ViewPager
    DisplayMode = { Auto | ButtonPanel | EditableComboBox }
    DisplayModeStyle  { "" }
    NumberOfIndicesShowing = { # }
    LastButton = { Auto | AlwaysVisible | NotVisibleOnLast | None } 
    LastButtonGlyph = { "" }
    LastButtonText = { "" }
    LastButtonStyle = { "" }
    FirstButton = { Auto | AlwaysVisible | NotVisibleOnFirst | None }
    FirstButtonGlyph = { "" }
    FirstButtonText = { "" }
    FirstButtonStyle = { "" }
    NextButton = { Auto | AlwaysVisible | NotVisibleOnLast | None } 
    NextButtonGlyph = { "" }
    NextButtonText = { "" }
    NextButtonStyle = { "" }
    PreviousButton = { Auto | AlwaysVisible | NotVisibleOnFirst | None }  
    PreviousButtonGlyph = { "" }
    PreviousButtonText = { "" }
    PreviousButtonStyle = { "" }
    EllipsisEnabled = { True | False }
    EllipsisMaxBefore = { # }
    EllipsisMaxAfter = { # }
    EllipsisShowBounds = { True | False }
    PrefixText = { "" }  
    NumberOfPages = { # } >
</ViewPager>
```

## Open Questions

* What should this control be named? 
* Should numerals in the button panel be replacable with glyphs? Will some scenarios require the glyphs to not be clickable (buttons not needed and spacing may be narrowed)? Should a timer and wrap-around feature be added?

**Comment by ryandemopoulos** [comment]
Made a few minor updates Savoy:

- Added a hyperlink to the WCT DataGrid control, since it's part of a separate library
- Updated the ""s and #s to be string and int respectively, in the code snippet
- Replaced ViewPager with PagerControl, to match the title.

**Comment by kmgallahan** [comment]
Great proposal. 

I feel like with this display mode:

![Numeric Display Mode](https://user-images.githubusercontent.com/16964652/49760354-23671280-fc79-11e8-804b-71b5b84a80f5.png)

The First & Last buttons are redundant, so the default state could be None.

Regarding these open questions:

> Should numerals in the button panel be replaceable with glyphs? 

Seems to be a popular design choice, and I'd like to have the option.

> Will some scenarios require the glyphs to not be clickable (buttons not needed and spacing may be narrowed)? 

- I'm sure some designers might want this option, but as a user I pretty much always dislike when I can't interact with them.

Should a timer and wrap-around feature be added?

- Would be useful, but low priority IMO.

**Comment by michael-hawker** [comment]
I think it's important to highlight the 'not in scope' part in the rationale a bit better, as that was my first question/thought:

> The scope of this proposal is not provide a data paging solution at this time but to start with UI paging as a foundation on top of which data paging may be integrated.

So, to clarify, this is a UI element only, but the developer is still responsible for manipulating the source to the GridView/ListView/etc... to make the pages work?  It'd be good to provide a code sample here about how that's expected to be accomplished to see how this scenario would work end-to-end.

(Also, in the sample XAML, don't indent the pager control as it makes it look like it goes in the GridView component not next to it in the tree.)

**Comment by SavoySchuler** [comment]
@kmgallahan Great feedback, thank you! I agree it reduces redundancy to only have the LastButton/FirstButton or the range bounds visible. Upon reviewing several example implementations, it appears most actually have neither - which I will update the default behavior to be. Unless it causes confusion, I will leave the redundancy of these optional components in the Usage Examples to help communicate what they are and how they look. 

I agree that glyph panels are a popular design. I am not sure if I would if it would make more sense to have "GlyphPanel" as a DisplayMode, subclass of a pager parent, or just as an option to replace numerals with glyphs in the ButtonPanel DisplayMode. When you think of wanting to create a glyph panel for these kind of displays, are one of these options where you would think to start?

**Comment by kmgallahan** [comment]
@SavoySchuler My initial thought was that a subclass would somewhat defeat the purpose of the `DisplayMode` property. However, looking over the properties & proposed features, it seems many of them would only be applicable to a particular display mode:

**Comment by SavoySchuler** [comment]
@kmgallahan I agreed! The button panel and combobox scenarios certainly have more in common with each other than either of them do with glyphs. Consider a pager base class with two subclasses: one for glyphs and another for both ButtonPanel + ComboBox (which uses the display mode property to set the core indexing component). Does this feel like an ideal compromise to pair the two display modes with common properties as one sublass and to share all the methods and events common to all three from the base class? I feel like this the reduces redundancy of a sublcass-only approach and adds clarity by removing atavistic properties left by the DisplayMode only approach, but I also want to make sure I am not adding complexity with out benefit.

**Comment by kmgallahan** [comment]
@SavoySchuler Seems like a good compromise yes. Having the base class also provides an opportunity for reuse in the future, if there were ever some 4th/5th modes that one might want to introduce (without having to always propose more `DisplayMode`s).

**Comment by SavoySchuler** [comment]
@betrue-final-final Thank you for this feedback! I will set the default display mode to be ComboBox. 

Regarding the glyphic display mode, do you think it would it be an appropriate compromise for it to visually display pips but read out numerals to narrator? In this case, you could think of them as the same display mode that differs only in whether glyphs or numerals are visually displayed on the button panel but all other accessiblity navigation and output would be the same. This glyphic display pops up in a lot of store applications and are often custom implementations that are not accessibility inclusive. I believe we have an opportunity to create an accessible version of this control where there have previously been few. Please let me know if you see things differently as inclusion of the glyphic display mode is still an open question. Thanks again for starting this conversation!

**Comment by micahl** [comment]
An additional open question I think would be great to have answered from the community is whether anyone has encountered situations other than a FlipView UX that required the use of the dots / glyphs?  The [FlipView's design guidance](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/flipview) refers to it as a type of context indicator.  If it's only used with a FlipView UX then should FlipView just provide that as a built-in option?

**Comment by SavoySchuler** [comment]
@micahl Great question! I have updated the Open Questions section to include this.

**Comment by LucasHaines** [comment]
Is this something we could use for all Forms? See the last comment in Issue #82.

**Comment by mrlacey** [comment]
How will this control react to a lack of space for all sub-elements having space to be displayed? -- Will the number of options displayed be reduced? Will the control be clipped if it overflows the available space?
Will the space between sub-elements be further padded when there is additional space to expand into?

**Comment by mdtauk** [comment]
I think when talking about the forms control, a pagination control of this complexity would not be ideal, and could encourage complexity where it need not be.

When I commented on pagination I was thinking more about the Xbox or Windows OOBE style of pagination or grouping of steps - so you see how far into the form process you are.

![image](https://user-images.githubusercontent.com/7389110/50180514-b64f1f00-0301-11e9-90f9-b75c44cd3e27.png)

![image](https://user-images.githubusercontent.com/7389110/50180564-e0084600-0301-11e9-8081-de7ad11f37a7.png)

Maybe this type of simple pagination can be part of the control proposed - perhaps it could be seen as a primitive used by various controls - but pagination as a form of progress indication.

**Comment by SavoySchuler** [comment]
@LucasHaines, @mrlacey is correct that this control would accomodate your scenario and the core functionality of the ones surfaced by by @mdtauk (noting we would need to expand on the visual styling to meet these scenarios fully). 

@mrlacey Your question is one that I am actively exploring. I am starting to believe that offering this widget without built-in data paging capabilities will not be enough to fully enable the experiences we would like to help developers create for their customers. I am thinking that this proposal might benefit from being expanded to include an optional data paging component that would handle data paging, displaying content, and window resizing. Leaving this component optional would also still allow for more advanced and custom scenarios that make use of only the UI component like @LucasHaines might need for forms or @mdtauk might need for grouping set up steps.

**Comment by SavoySchuler** [comment]
@mrlacey Thanks for your feedback on the glyphic display mode! Do you happen to have any of those examples of brands using custom glyphs that you can point me toward?

**Comment by micahgodbolt** [comment]
Hey everyone. As we're working on the pagination control over in Fabric, we were considering building a more primitive "pager" control that is mostly this:

![image](https://user-images.githubusercontent.com/1434956/50708135-16d79f00-1018-11e9-9657-ae7508daa3ff.png)

It'd maintain state, have first/prev/next/last controls, have a public interface for changing state etc, but wouldn't be concerned with how the paging view was represented (numbers, numbers with ellipsis, glyphs, or even pivot buttons). It would also include an optional child view (above or below) to render content based on current page, making it foundation for carousels, FREs, pivots etc. 

Do you typically do similar abstractions in UWP? Is this something you're considering with this control?

**Comment by mrlacey** [comment]
@SavoySchuler  Sorry, nothing public I can point to anymore. Maybe it's a trend that's passed now but I have memories of re-templating lots of low-level things like a progress ring so it used different shapes, instead of circles, etc. 
When this issue came up I immediately thought of designers I've worked with in the past who would love the ability to easily change these glyphs.

**Comment by mdtauk** [comment]
If the glyphs for each button are exposed as a DependencyProperty with default values set for the SegoeMDL2 fonts - then creating a style with new FontFamily and Glyph values set should provide the flexibility and ease of change devs who need it, would want.

**Comment by YuliKl** [comment]
> It would also include an optional child view (above or below) to render content based on current page, making it foundation for carousels, FREs, pivots etc.
> Do you typically do similar abstractions in UWP?

@micahgodbolt - No, UWP controls tend to be complete and ready to use out of the box rather than being smaller building blocks. Molecules rather than atoms, if that analogy makes sense. We have a Pivot control which is in no way related to the proposed PagerControl.

**Comment by micahgodbolt** [comment]
@YuliKl totally understand. In React we're always looking for ways to reuse smaller controls, or to breaking up a single control into smaller reusable pieces. As i'm less familiar with UWP, I'm not sure how well the platform supports that type of sharing.

**Comment by micahl** [comment]
I think you'd find in practice that there's more similarity than differences.  Bigger controls (molecules) that contain a lot of built-in policy are made up of smaller controls (atoms) that have less policy.  Molecules are easy to use.  Atoms are more flexible and simple.  Expectations around a given scenario and its similarity to others heavily influence the approach taken to solve it.  @micahgodbolt It would be useful to understand what scenarios you've encountered.

**Comment by micahgodbolt** [comment]
@micahl When I look at pager controls I see three things, 1 stateful container with controls to go to first, prev, next last, and secondly a view to display the current page and give UI for selecting the page, and lastly an optional content region to render content associated with that page. So if we built a paging control that only worried about state, the inner and outer view, and controls/public methods for changing the page, then we'd have a really flexible control that could be used to create carousels, first run experiences, pivots, and pagination.

This way everything that was built on this control would share a common public interface, and could easily be customized to add functionality, or change the way in which the UI is displayed. It also means that when we go to add functionality, we can add it in one place and everything else gains the same functionality. We already have too many similar controls where a feature is added to one of them, and then someone asks for that feature to be added to others. 

Great example, we have an underlined TextField variant (just an underline, no box border), and there's a request to add that variant to datepicker, dropdown and comobobox (it's already added to search). https://github.com/OfficeDev/office-ui-fabric-react/issues/5631

So you can see it would have been better for all of those controls to be built off of TextField, so that they could each benefit from that design variant.

**Comment by mdtauk** [comment]
Fabric and FluentWeb are all seemingly isolated Web implementations of the Fluent Design ideals.

The Windows XAML controls were separate and part of the OS until Windows UI Library. 

So as the Windows UI Library slowly takes over as the default way to handle XAML controls, maybe the XAML control templates can slowly adopt the alternative styles and other controls to match the Web implementations.  

The CommandBarFlyout, DropDownButton, etc which were added in 1803-1809 were the result of the Office UWP team folding back work they did for all devs to use.  So lets keep our fingers crossed

**Comment by mdtauk** [comment]
> > It would also include an optional child view (above or below) to render content based on current page, making it foundation for carousels, FREs, pivots etc.
> > Do you typically do similar abstractions in UWP?
> 
> @micahgodbolt - No, UWP controls tend to be complete and ready to use out of the box rather than being smaller building blocks. Molecules rather than atoms, if that analogy makes sense. We have a Pivot control which is in no way related to the proposed PagerControl.

There is a Primitives namespace, and I there are controls like the NavigationView which creates its own sub controls like NavigationViewItem - or Pivot and its PivotHeaderItem, PivotViewItem.

So it is perfectly possible to create sub controls designed to be consumed and used by a "full control" which a dev places onto the design view.

**Comment by michael-hawker** [comment]
Yeah, I don't think XAML controls have to be all encompassing.  In fact I feel there can be cases where large controls should be made out of smaller controls which could be consumed elsewhere as well.

For instance part of a NavigationView could be a TabView control.  And the InkToolbar is made up of various Buttons and things.

So I think there's places and patterns for both.

In terms of a pager control, I could see the 'visualization' of numbers or dots be a some other visualization component that could be restyled or such vs. the pieces that control the actual navigation.

**Comment by micahgodbolt** [comment]
Having done this a few years, and worked with a lot of partner teams using our framework, I'm quite familiar with all of the customization requests that we get. "I want to use your control, but swap in this piece for that piece". So we really try to create controls with that functionality in mind. That can be best done by creating the controls via composition of many smaller, single function pieces.

**Comment by SavoySchuler** [comment]
@micahgodbolt Thank you for surfacing this conversation! The intention for this control would be to develop it in the way you are suggesting where we would have a basic parent class that would offer states, events, and maybe first/back/next/last buttons from which we could subclass a numerical button panel pager, a glyphic pager, and ComboBox pager. This will allow developers to leverage the states and events class while still having the option for a free range of customization with the core pager control.

**Comment by micahgodbolt** [comment]
This is exactly why I included imports/exports into our new control proposal issue. I recognize that very few controls we build are a single control, and that most are composed of several others which may or may not have yet been created. 

https://github.com/OfficeDev/office-ui-fabric-react/blob/master/.github/ISSUE_TEMPLATE/new_component.md#exportscomponent-breakdown

**Comment by SavoySchuler** [comment]
Agreed. Thanks again, @micahgodbolt! I have updated our Requirements section to reflect this componentization as a core priority for the control.

**Comment by mdtauk** [comment]
The DisplayMode property could be used to enable a simple step based representation, as used by Windows and XBox OOBE sequences - and could be useful for onboarding scenarios for app developers.

![image](https://user-images.githubusercontent.com/7389110/51570129-29fec180-1e96-11e9-8ae0-bdf314b03f7d.png)

![image](https://user-images.githubusercontent.com/7389110/51570142-33882980-1e96-11e9-8733-c97f313a251c.png)

You could even allow it to be template-able, but as a default appearance, something like this would fit with UWP patterns.

![image](https://user-images.githubusercontent.com/7389110/51571255-8f07e680-1e99-11e9-825b-036f16ddc886.png)

There could be properties to hide or show the labels, or it could be part of the template.

**Comment by SavoySchuler** [comment]
@mdtauk Exactly, under the current proposal, your scenario is one that would be enabled by a base parent class designed to handle states and events. 

As for including a sequence based display as you have suggested, I have added this to the Open Questions section to collect feedback on whether this would be a popular out-of-the-box addition for developers.

**Comment by mdtauk** [comment]
:) @SavoySchuler That primitive control could also be used by a Form control that has been proposed, as well as for any future "Wizard" control

**Comment by SavoySchuler** [comment]
@mdtauk Nice catch! The [proposed Form](https://github.com/Microsoft/microsoft-ui-xaml/issues/82) control paired with a step based pager would make for a fantastically smooth experience creating those scenarios.

**Comment by SavoySchuler** [comment]
Thank you again for your feedback, everyone! In response to some earlier comments and to better support more traditional scenarios where this control can be used, I have posted a separate proposal for [adding a data paging component to PagerControl](https://github.com/Microsoft/microsoft-ui-xaml/issues/268). This addition would make PagerControl functionally comparable to Silverlight's DataPager with an improved user experience via dynamic resizing and state tracking.

These proposals are distinct because I believe a flexible UI paging widget can valuable in its own right - which I believe is supported by the similar developments of the Fabric team/@micahgodbolt. This means if the data paging add-on is something you would like to see for this control, please go comment or thumbs-up [here](https://github.com/Microsoft/microsoft-ui-xaml/issues/268)!

**Comment by mdtauk** [comment]
So a **Pagination** control, which includes **ProgressSteps**, and **Pager** controls which would be shared by a **DataPager** control, which could be an extended **DataGrid** control?

**Comment by SavoySchuler** [comment]
Hi, @mdtauk! You're right, this is getting a little messy. Here's a breakdown of everything proposed here and in [Proposal #268](https://github.com/Microsoft/microsoft-ui-xaml/issues/268) with the unique features as sub-bullets.

Base Class: 

- PagerControl
    - Events
    - States
    - Back/Next Buttons

Sub Classes:

- ButtonPanel
    - First/Last Buttons
- ComboBox
    - First/Last Buttons
- Glyphic  [Proposed here]
    - Wrap
    - Timer
- ProgressSteps  [Proposed here]
    - Text labels

Sister class to PagerControl:

- DataPager (data side) [[Proposal #268]](https://github.com/Microsoft/microsoft-ui-xaml/issues/268)
    - Data paging and serving
    - Dynamic page resizing
    - State management

**Comment by SavoySchuler** [comment]
@mdtauk, I looked into the ProgressSteps scenario (commonly called "First Run Experience" or FRE) more and I am wondering if you think the [Pivot Control](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/pivot) paired with the PagerControl base class would adequately solve the FRE scenario for you? Pivot can have its tabs programmatically enabled/disabled and the navigation events between Pivot and PagerControl could be synced. Is there anything Pivot is missing that you would need?

**Comment by mdtauk** [comment]
In the Windows and Xbox OOBEs that use that type of paging/progress - there is no swiping to go back and forth, instead they have a back button and a Next command button.

I think it may need to be an inactive control to show the progress, but also to keep track of the inputs entered in previous steps, so you can go back and change your inputs.

I see it more as a basic readout of the "pages" in a linear fashion, so moving forward through pages, equals progressing through steps.

**Comment by michael-hawker** [comment]
Related to this [UserVoice item](https://wpdev.uservoice.com/forums/110705-universal-windows-platform/suggestions/36536572-circular-flipview-control-with-position-indicators) as well.

**Comment by SavoySchuler** [comment]
@mdtauk and @michael-hawker, I met with my team today and we scoped this development down to focus on the core pager scenarios which are now shown in the summary. 

@mdtauk, would you please open a new proposal for the ProgressSteps/FRE scenario you brought up? I am happy to do it, but I would prefer we capture this scenario fully in your own words. Plus, this way ensures you get due credit! I am happy to fill out any remaining details.

**Comment by SavoySchuler** [comment]
All, I have opened a new proposal for [glyphic index displays associated with FlipView](https://github.com/Microsoft/microsoft-ui-xaml/issues/602). Please thumbs up this proposal separately if you are interested in seeing this feature added to FlipView. Otherwise, this proposal has been approved for spec writing and I will be adding links to a spec branch shortly.

**Comment by SavoySchuler** [comment]
# This proposal has been approved for [spec writing](https://github.com/Microsoft/microsoft-ui-xaml-specs/blob/user/savoyschuler/pagercontrol/active/PagerControl/PagerControl.md).

Please feel encouraged to continue high-level discussion here on the proposal. More detailed discussion will occur on the Spec and the associated PR once it is available.

**Comment by mdtauk** [comment]
Will there be support for pages of content without data binding?

Something like:
```xaml
<PagerControl>
      <PagerControl.Page>
            <TextBlock Text="Section 1" />
            <Button Content="Next" />
      </PagerControl.Page>
      <PagerControl.Page>
            <TextBlock Text="Section 2" />
            <Button Content="Cancel" />
            <Button Content="Confirm" Style="{ThemeResource AccentButton}" />
      </PagerControl.Page>
</PagerControl>
```

It would be useful for Custom Controls like signup or login forms, which could be used throughout an app

**Comment by SavoySchuler** [comment]
> Will there be support for pages of content without data binding?

This is a must (I have heard the same ask from internal customers) and I have updated the Requirements list to reflect this. My current understanding is that the API can (mostly) look the same as it will in solutions that leverage data binding, the distinguishing work will need to be handled in the navigation events that would require some special casing to handle navigation calls to content that may or may not exist + possible logic to update the UI if a determinate page count is established.

**Comment by michael-hawker** [comment]
@mdtauk @SavoySchuler correct me if I'm wrong, but the PagerControl isn't going to own the content it's paging through, correct? I think hosting content is out-of-scope for the purpose of the PagerControl?

I noticed in the spec there's a lot of commands for first/next/prev/last... it'd be cool if a control could implement an interface and hook itself in to the Pager somehow? (Thinking of `TargetInkCanvas` type model from `InkToolbar`.) The interface would provide details to the pager on the number of items/pages/current page and can be called upon to move to the page requested by the pager control on user interaction?

Then it'd be helpful for certain controls like `FlipView` to implement this interface (in WinUI 3) so that these scenarios can just work together automatically for some of the standard UX experiences (like the [context indicator](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/flipview#adding-a-context-indicator) called out as a sample currently here would become much easier).

```xml
     <muxc:FlipView x:Name="itemFlipView" ItemsSource="{Binding Source={StaticResource itemsViewSource}}"/>
     <muxc:PagerControl TargetControl="{Binding itemFlipView}"/>
```

**Comment by michael-hawker** [comment]
Also, I didn't notice an `Orientation` property called out, I would think that you could want a `PagerControl` on the side as well as top/bottom?

**Comment by YuliKl** [comment]
> Also, I didn't notice an `Orientation` property called out, I would think that you could want a `PagerControl` on the side as well as top/bottom?
> 
> Thinking of a case of having a side-by-side page preview but with the pager in the middle (or on the sides):

This feels like a pretty different scenario to me... and I don't think the design takes this scenario into account.

**Comment by MarissaMatt** [comment]
@michael-hawker yes you are correct the pager control isn't going to own the content it's paging through. It will just be a UI component that will have an event to notify the developer which page should be displayed. For your point about having a control implement an interface that can hook in to the pager control are you thinking of something like this proposal (https://github.com/Microsoft/microsoft-ui-xaml/issues/268)? 

@YuliKl you are correct the design has not considered a scenario where the pager could have a different orientation and it outside of the scope for this control. Mainly because the majority of the use cases and scenarios we are going to cover only show the pager below or above a layout view.

**Comment by StephenLPeters** [comment]
@chingucoding can you comment on this issue so that I can assign it to you?

**Comment by chingucoding** [comment]
Sure thing.

**Comment by michael-hawker** [comment]
@YuliKl @MarissaMatt yeah, different scenarios and use-cases, but still can be covered by this same control/pattern.

I think it's important to consider alternatives when designing controls to make sure they can provide for a broad range of experiences and ensure those experiences are consistent across apps.

I've only seen the pager in grid like scenarios in the mock-up, but I see driving page navigation in reading apps or document editors as an equally valid use-case.

**Comment by gabbybilka** [comment]
For anyone interested in pip pagination (oriented vertically and horizontally) leveraging PagerControl, please continue the discussion in this recently opened issue #3464. Thanks!

**Comment by chingucoding** [comment]
Are there plans to also provide this control in WinUI 3?

**Comment by jingliancui** [comment]
Hi All, do you know is this spec will plan for winui3? Or any others 3rd party nuget we can use for winui3? Thanks.

**Comment by marcelwgn** [comment]
@jingliancui That would be something I could work on. Are you using C# or C++?

**Comment by jingliancui** [comment]
Hi @marcelwgn , thanks your kindly response,  I'm using C# MAUI and WinUI3.

**Comment by marcelwgn** [comment]
Thank y @jingliancui, so would a WinUI 3 only solution help you then?

**Comment by jingliancui** [comment]
Thanks @marcelwgn is it easy for you to import it to winui3? If yes then please.Thanks

**Comment by marcelwgn** [comment]
I'll give it a shot to write a lib that only contains PagerControl so it at least can be imported when you need it @jingliancui

**Comment by jingliancui** [comment]
If you are get familiar with maui then it would be great that the ui can be shared to Mac, Win,Android,iOS, iPadOS.

**Comment by jingliancui** [comment]
Thank you so much, if Winui3 take you fewest effort, then winui3 is ok. Thanks again

**Comment by jingliancui** [comment]
Hi @marcelwgn Do you have the repo link to share? So I can try to help if need. Thanks.

**Comment by marcelwgn** [comment]
Hi @jingliancui thank you for asking, I'm afraid to say I don't have a repro (or NuGet) link yet but I'll be sure to let you know once I have something. Thank you for your interest in this!

**Comment by marcelwgn** [comment]
Hi @jingliancui, I've ported it over to C# and published the control to NuGet now. You can find the repo here: https://github.com/marcelwgn/PagerControlPolyfill

**Comment by jingliancui** [comment]
Thank you so much @marcelwgn !
